File: pi42/market.py
"""
Market data API endpoints for Pi42.
"""

from typing import Dict, Any, Optional, List, Union


class MarketAPI:
    """
    Market API client for accessing Pi42 market data endpoints.
    """

    def __init__(self, client):
        """
        Initialize the Market API client.

        Args:
            client: The main Pi42Client instance
        """
        self._client = client

    def get_ticker_24hr(self, contract_pair: str) -> Dict[str, Any]:
        """
        Get 24-hour ticker data for a specific trading pair.

        Args:
            contract_pair: Trading pair symbol (e.g., 'btc' for BTCINR)

        Returns:
            Dict: 24-hour ticker data
        """
        endpoint = f"/v1/market/ticker24Hr/{contract_pair.lower()}"
        return self._client.get_request(endpoint, public=True)

    def get_agg_trades(self, contract_pair: str) -> Dict[str, Any]:
        """
        Get aggregated trade data for a specific trading pair.

        Args:
            contract_pair: Trading pair symbol (e.g., 'btc' for BTCINR)

        Returns:
            Dict: Aggregated trade data
        """
        endpoint = f"/v1/market/aggTrade/{contract_pair.lower()}"
        return self._client.get_request(endpoint, public=True)

    def get_depth(self, contract_pair: str) -> Dict[str, Any]:
        """
        Get order book depth data for a specific trading pair.

        Args:
            contract_pair: Trading pair symbol (e.g., 'btc' for BTCINR)

        Returns:
            Dict: Order book depth data
        """
        endpoint = f"/v1/market/depth/{contract_pair.lower()}"
        return self._client.get_request(endpoint, public=True)

    def get_klines(self, pair: str, interval: str, 
                  start_time: Optional[int] = None, 
                  end_time: Optional[int] = None, 
                  limit: Optional[int] = None) -> List[Dict[str, Any]]:
        """
        Get candlestick (kline) data for a specific trading pair and interval.

        Args:
            pair: Trading pair symbol (e.g., 'BTCINR')
            interval: Kline interval (e.g., '1m', '5m', '1h')
            start_time: Start time in milliseconds
            end_time: End time in milliseconds
            limit: Maximum number of klines to return

        Returns:
            List[Dict]: List of kline data
        """
        endpoint = "/v1/market/klines"
        
        params = {
            "pair": pair.upper(),
            "interval": interval.lower()
        }
        
        if start_time:
            params["startTime"] = start_time
        if end_time:
            params["endTime"] = end_time
        if limit:
            params["limit"] = limit
            
        return self._client.post_request(endpoint, params, public=True)

File: market.go
package pi42

import (
	"encoding/json"
	"fmt"
	"strings"
)

// MarketAPI provides access to market data endpoints
type MarketAPI struct {
	client *Client
}

// NewMarketAPI creates a new Market API instance
func NewMarketAPI(client *Client) *MarketAPI {
	return &MarketAPI{client: client}
}

// GetTicker24hr gets 24-hour ticker data for a specific trading pair
func (api *MarketAPI) GetTicker24hr(contractPair string) (map[string]interface{}, error) {
	endpoint := fmt.Sprintf("/v1/market/ticker24Hr/%s", strings.ToLower(contractPair))
	params := make(map[string]string)

	data, err := api.client.Get(endpoint, params, true)
	if err != nil {
		return nil, err
	}

	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("error parsing response: %v", err)
	}

	return result, nil
}

// GetAggTrades gets aggregated trade data for a specific trading pair
func (api *MarketAPI) GetAggTrades(contractPair string) (map[string]interface{}, error) {
	endpoint := fmt.Sprintf("/v1/market/aggTrade/%s", strings.ToLower(contractPair))

	data, err := api.client.Get(endpoint, nil, true)
	if err != nil {
		return nil, err
	}

	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("error parsing response: %v", err)
	}

	return result, nil
}

// GetDepth gets order book depth data for a specific trading pair
// Returns structured DepthResponse containing order book bids and asks
func (api *MarketAPI) GetDepth(contractPair string) (*DepthResponse, error) {
	endpoint := fmt.Sprintf("/v1/market/depth/%s", strings.ToLower(contractPair))

	data, err := api.client.Get(endpoint, nil, true)
	if err != nil {
		return nil, err
	}

	var result DepthResponse
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("error parsing depth response: %v", err)
	}

	return &result, nil
}

// KlinesParams represents parameters for the Klines method
type KlinesParams struct {
	Pair      string `json:"pair"`                // Trading pair (e.g., "BTCINR")
	Interval  string `json:"interval"`            // Kline interval (e.g., "1m", "1h", "1d")
	StartTime int64  `json:"startTime,omitempty"` // Optional start time in milliseconds
	EndTime   int64  `json:"endTime,omitempty"`   // Optional end time in milliseconds
	Limit     int    `json:"limit,omitempty"`     // Optional limit on number of results
}

// GetKlines gets candlestick (kline) data for a specific trading pair and interval
// Returns an array of structured KlineData objects
func (api *MarketAPI) GetKlines(params KlinesParams) ([]KlineData, error) {
	endpoint := "/v1/market/klines"

	// Convert struct to map for the request
	paramsMap := map[string]interface{}{
		"pair":     strings.ToUpper(params.Pair),
		"interval": strings.ToLower(params.Interval),
	}

	if params.StartTime > 0 {
		paramsMap["startTime"] = params.StartTime
	}
	if params.EndTime > 0 {
		paramsMap["endTime"] = params.EndTime
	}
	if params.Limit > 0 {
		paramsMap["limit"] = params.Limit
	}

	data, err := api.client.Post(endpoint, paramsMap, true)
	if err != nil {
		return nil, err
	}

	var result []KlineData
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("error parsing klines response: %v", err)
	}

	return result, nil
}

// For backward compatibility
func (api *MarketAPI) Ticker24Hr(contractPair string) (map[string]interface{}, error) {
	return api.GetTicker24hr(contractPair)
}